daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -860,47 +860,52 @@ private[lf] final class Compiler(
       tmpl: Template,
   ): (t.SDefinitionRef, SDefinition) =
     unlabelledTopLevelFunction2(t.ToContractInfoDefRef(tmplId)) { (tmplArgPos, mbKeyPos, env) =>
-      // We use a chain of let bindings to make the evaluation order of SBuildContractInfoStruct's arguments is
-      // independent from the evaluation strategy imposed by the ANF transformation.
-      let(env, s.SEValue(STypeRep(TTyCon(tmplId)))) { (typePos, env) =>
-        let(env, t.AgreementTextDefRef(tmplId)(env.toSEVar(tmplArgPos))) {
-          (agreementTextPos, env) =>
-            let(env, t.SignatoriesDefRef(tmplId)(env.toSEVar(tmplArgPos))) {
-              (signatoriesPos, env) =>
-                let(env, t.ObserversDefRef(tmplId)(env.toSEVar(tmplArgPos))) {
-                  (observersPos, env) =>
-                    val body = tmpl.key match {
-                      case None =>
-                        s.SEValue.None
-                      case Some(tmplKey) =>
-                        s.SECase(
-                          env.toSEVar(mbKeyPos),
-                          List(
-                            s.SCaseAlt(
-                              t.SCPNone,
-                              let(
-                                env,
-                                translateExp(env.bindExprVar(tmpl.param, tmplArgPos), tmplKey.body),
-                              ) { (keyPos, env) =>
-                                SBSome(translateKeyWithMaintainers(env, keyPos, tmplKey))
-                              },
+      checkPreCondition(env, tmplId, env.toSEVar(tmplArgPos)) { env =>
+        // We use a chain of let bindings to make the evaluation order of SBuildContractInfoStruct's arguments is
+        // independent from the evaluation strategy imposed by the ANF transformation.
```

Comment:
Maybe a good idea to put the `checkPreCondition`, bellow the comment as is is basically a let binding